executor/aggfuncs/func_max_min.go

```diff
@@ -17,13 +17,145 @@ import (
+}
+
+// Deque is an array based double end queue
+type Deque struct {
+	Items   []Pair
+	IsMax   bool
+	cmpFunc func(i, j interface{}) int
+}
+
+// PushFront pushes idx and item(wrapped in Pair) to the front of Deque
+func (d *Deque) PushFront(idx uint64, item interface{}) {
+	p := Pair{item, idx}
+	d.Items = append([]Pair{p}, d.Items...)
+}
+
+// PushBack pushes idx and item(wrapped in Pair) to the end of Deque
+func (d *Deque) PushBack(idx uint64, item interface{}) {
+	p := Pair{item, idx}
+	d.Items = append(d.Items, p)
+}
+
+// PopFront pops an item from the front of Deque
+func (d *Deque) PopFront() (interface{}, error) {
+	if len(d.Items) <= 0 {
+		return nil, errors.New("Pop front when deque is empty")
+	}
+	defer func() {
+		d.Items = d.Items[1:]
+	}()
+	return d.Items[0], nil
+}
+
+// PopBack pops an item from the end of Deque
+func (d *Deque) PopBack() (interface{}, error) {
+	i := len(d.Items) - 1
+	if i < 0 {
+		return nil, errors.New("Pop back when deque is empty")
+	}
+	defer func() {
+		d.Items = d.Items[:i]
+	}()
+	return d.Items[i], nil
+}
+
+// Back returns the element at the end of Deque
+func (d *Deque) Back() (Pair, error) {
+	i := len(d.Items) - 1
+	if i < 0 {
+		return Pair{}, errors.New("Cannot get back when deque is empty")
+	}
+	return d.Items[i], nil
+}
+
+// Front returns the element at the front of Deque
+func (d *Deque) Front() (Pair, error) {
+	if len(d.Items) <= 0 {
+		return Pair{}, errors.New("Cannot get back when deque is empty")
+	}
+	return d.Items[0], nil
+}
+
+// IsEmpty returns if Deque is empty
+func (d *Deque) IsEmpty() bool {
+	if len(d.Items) == 0 {
+		return true
+	}
+	return false
+}
+
+// Pair pairs items and their indices in deque
+type Pair struct {
+	item interface{}
+	idx  uint64
+}
+
+// Reset deque
+func (d *Deque) Reset() {
+	d.Items = d.Items[:0]
+}
+
+// Dequeue pops out element from the front, if element's index is out of boundary, i.e. the leftmost element index
+func (d *Deque) Dequeue(boundary uint64) error {
+	for !d.IsEmpty() {
+		frontEle, err := d.Front()
+		if err != nil {
+			return err
+		}
+		if frontEle.idx <= boundary {
+			_, err := d.PopFront()
+			if err != nil {
+				return err
+			}
+			continue
+		}
+		return nil
+	}
+	return nil
+}
+
+// Enqueue put item at the back of queue, while popping any element that is lesser element in queue
+func (d *Deque) Enqueue(idx uint64, item interface{}) error {
+	if d.IsEmpty() {
+		d.PushBack(idx, item)
+		return nil
+	}
+
+	for !d.IsEmpty() {
+		pair, err := d.Back()
+		if err != nil {
+			return err
+		}
+
+		bigger := d.cmpFunc(item, pair.item) >= 0
+		// 1. if deque aims for finding max and item is bigger than element at back
+		// 2. if deque aims for finding min and item is smaller than element at back
+		if bigger && d.IsMax || !bigger && !d.IsMax {
+			_, err := d.PopBack()
+			if err != nil {
+				return nil
+			}
+		} else {
+			break
+		}
+	}
+	d.PushBack(idx, item)
+	return nil
+}
+
 type maxMinHeap struct {
```

Comment:
The `maxMinHeap` is no usages any more, Should we delete it?

Reply:
Thanks for reminding, checked it's safe to delete.